packages/flutter/lib/src/rendering/sliver_persistent_header.dart

```diff
@@ -280,6 +348,40 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
     return math.min(0.0, paintExtent - childExtent);
   }
 
+  /// If the header isn't already fully exposed, then scroll it into view.
+  void maybeStartSnapAnimation(ScrollDirection direction) {
+    if (snapConfiguration == null)
+      return;
+    if (direction == ScrollDirection.forward && _effectiveScrollOffset <= 0.0)
+      return;
+    if (direction == ScrollDirection.reverse && _effectiveScrollOffset >= maxExtent)
+      return;
+
+    final TickerProvider vsync = snapConfiguration.vsync;
+    final Duration duration = snapConfiguration.duration;
+    _controller ??= new AnimationController(vsync: vsync, duration: duration)
+      ..addListener(() {
+        if (_effectiveScrollOffset == _animation.value)
+          return;
+        _effectiveScrollOffset = _animation.value;
```

Comment:
It's weird that _animation has to be a class member only because that way it gets captured correctly here and so you don't have to re-add the listener the second time you animate even though you recreate the animation.